lib/Mitigation.pm

```diff
@@ -0,0 +1,352 @@
+    );
+    return $edx;
+}
+
+
+sub show {
+    my $self = shift;
+    print $self->Name(),  ",";
+    print $self->CPUID(), ",";
+    print $self->MSR(),   "\n";
+}
+
+sub vulnerabilities {
+    my $self = shift;
+    if ($self->read_cpuid() & $self->CPUID()) {
+        if ($self->read_msr() & $self->MSR()) {
+            record_info("Not Affected", "This machine needn't be test.");
+            return 0;    #Not Affected
+        } else {
+            record_info("vulnerable", "This machine need to test.");
+            return 1;    #Affected
+        }
+    }
+    record_info("vulnerable", "This machine need to test.");
+    return 1;
+}
+
+sub sysfs {
+    my ($self, $value) = @_;
+    my $p;
+    if (@_ == 2) {
+        return $self->{'sysfs'}->{$value};
+    }
+    return $self->{'sysfs'};
+
+}
+
+sub dmesg {
+    my $self = shift;
+    my $p;
+    for $p (keys %{$self->{'dmesg'}}) {
+        print "dmesg ", $self->Name, "\n";
+        print $self->{'dmesg'}->{$p}, "\n";
+    }
+}
+
+sub cmdline {
+    my $self = shift;
+    my $p;
+    return $self->{'cmdline'};
+}
+
+sub lscpu {
+    my $self = shift;
+    my $p;
+    for $p (keys %{$self->{'lscpu'}}) {
+        print $p, "\n";
+    }
+}
+
+
+
+#This function will finish testing in default status.
+#As out of box testing. and clean up all mitigations parameters.
+sub check_default_status {
+    my $self = shift;
+    assert_script_run('cat /proc/cmdline');
+    my $ret  = script_run('grep -v "' . $self->{'parameter'} . '=[a-z]*" /proc/cmdline');
+    my $ret1 = script_run('grep -v "' . "mitigations" . '=[a-z]*" /proc/cmdline');
+    if ($ret ne 0 or $ret1 ne 0) {
+        remove_grub_cmdline_settings($self->{'parameter'} . "=[a-z]*");
+        remove_grub_cmdline_settings("mitigations=[a-z]*");
+        reboot_and_wait($self, 150);
+        assert_script_run('grep -v "' . $self->{'parameter'} . '=off" /proc/cmdline');
+    }
+}
+
+#Check cpu flags exist or not.
+#when $cmd is off, the judge is reverse.
+sub check_cpu_flags {
+    my $self = shift;
+    my $cmd  = shift;
+    my $flag;
+    assert_script_run('cat /proc/cpuinfo');
+    foreach $flag (@{$self->{'cpuflags'}}) {
+        #switch off a feature, it wouldn't be display in 'flags'
+        if ($cmd eq "off") {
+            assert_script_run('cat /proc/cpuinfo | grep -v "^flags.*' . $flag . '.*"');
+        } else {
+            assert_script_run('cat /proc/cpuinfo | grep "^flags.*' . $flag . '.*"');
+        }
+    }
+}
+
+sub check_sysfs {
+    my ($self, $value) = @_;
+    assert_script_run('cat ' . $syspath . $self->Sysfs());
+    if (@_ == 2) {
+        assert_script_run(
+            'cat ' . $syspath . $self->Sysfs() . '| grep ' . '"' . $self->sysfs($value) . '"');
+    }
+}
+
+sub check_dmesg {
+    my $self  = shift;
+    my $value = shift;    #the value of kernel parameter
+    my $string;
+
+    foreach $string ($self->{'dmesg'}->{$value}) {
```

Comment:
the iterator variable can directly be declared within the loop:

```perl
for my $string ($self->{dmesg}->{$value}) {
    assert_script_run(…)
}

Reply:
Got it.